### PR TITLE
Allow users to opt out of exception reporting with CYPRESS_CRASH_REPORTS=0

### LINKED ISSUE #4386: Opt out of sending exceptions to https://api.cypress.io/exceptions
When an exception is thrown, it appears that it is reported to `https://api.cypress.io/exceptions`

I noted this when running debug logs, and I was never aware that this occurred before. It should be possible to opt-out of sending this information, which is a standard setting in most applications.

```
cypress:network:agent addRequest called for https://api.cypress.io/exceptions +827ms
```

```
  cypress:server:api request to url: POST https://api.cypress.io/exceptions with params: {"body":{"err":{"name":"TypeError","message":"Cannot read property 'passes' of undefined","stack":"TypeError: Cannot read property 'passes' of undefined\n    at Spec.Base.epilogue (<stripped-path>base.js:319:26)\n    at Object.onceWrapper (events.js:315:30)\n    at emitOne (events.js:121:20)\n    at Runner.emit (events.js:211:7)\n    at Reporter.emit (<stripped-path>reporter.js:239:55)\n    at Object.server.startWebsockets.onMocha (<stripped-path>project.js:298:22)\n    at Socket.<anonymous> (<stripped-path>socket.js:235:36)\n    at emitTwo (events.js:126:13)\n    at Socket.emit (events.js:214:7)\n    at <stripped-path>socket.js:503:12\n    at _combinedTickCallback (internal<stripped-path>next_tick.js:131:7)\n    at process._tickCallback (internal<stripped-path>next_tick.js:180:9)\n"},"version":"3.3.3","osName":"darwin","osVersion":"18.6.0","osCpus":[{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":4240850,"nice":0,"sys":2171700,"idle":19296260,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":405060,"nice":0,"sys":275490,"idle":25025900,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":3776280,"nice":0,"sys":1593060,"idle":20337680,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":405460,"nice":0,"sys":245570,"idle":25055330,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":3391380,"nice":0,"sys":1305860,"idle":21009660,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":412440,"nice":0,"sys":223550,"idle":25070280,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":2899130,"nice":0,"sys":1034430,"idle":21773210,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":420540,"nice":0,"sys":203930,"idle":25081700,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":2506430,"nice":0,"sys":852440,"idle":22347800,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":426610,"nice":0,"sys":189060,"idle":25090410,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":2326920,"nice":0,"sys":745760,"idle":22633870,"irq":0}},{"model":"Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz","speed":2600,"times":{"user":429690,"nice":0,"sys":177820,"idle":25098480,"irq":0}}],"osMemory":{"free":217538560,"total":17179869184}},"headers":{"x-os-name":"darwin","x-cypress-version":"3.3.3"}}
```

### LINKED ISSUE #4500: Display `projectId` in the resolved list of configuration settings
To make `projectId` consistent with the rest of the config values - it'll now show up in your project settings.

image

### LINKED ISSUE #4501: Truncate special env vars and debug logs
Display them as `XXX...XXX`

### PR DESCRIPTION
**Pull request by flotwig**
- Fixes #4386 
- Fixes #4500
- Fixes #4501

* [x] If Cypress detects `CYPRESS_CRASH_REPORTS=0` is set in the environment, it will not send exceptions to `api.cypress.io` for debugging.

### CODE REVIEW

File: packages/server/lib/exception.coffee [hunk 1]
```diff
@@ -48,7 +48,9 @@ module.exports = {
       user and user.authToken
 
   create: (err) ->
-    return Promise.resolve() if process.env["CYPRESS_ENV"] isnt "production"
+    if process.env["CYPRESS_ENV"] isnt "production" or
```
Comment (Bkucera) [inline packages/server/lib/exception.coffee]: You can use Promise.method here http://bluebirdjs.com/docs/api/promise.method.html

**Comment by Bkucera** [comment]
We don't want this to be able to be set by Cypress config, it doesn't make sense to go there, so it's good that you use process.env directly. 

 It will be picked up by Cypress.env() so maybe we should remove it after finding it

**Comment by chrisbreiding** [comment]
Why not allow users to set it via `cypress.json`?

**Comment by brian-mann** [comment]
Because it's not project specific configuration that should be checked into source control - it's local preference behavior specific to the user. 

But more than anything else - Cypress could crash well before there is a `cypress.json`. It could crash at install time, or in global mode (before you add project). It could crash parsing `cypress.json` in fact.

An environment variable is the only way to guarantee that this option is set for the entirety of the application's lifecycle.

Also I'm thinking about renaming this to `CYPRESS_CRASH_REPORTS`.

**Comment by brian-mann** [comment]
Agree with @Bkucera 	this value should be sliced out of env.

**Comment by flotwig** [comment]
`CYPRESS_CRASH_REPORTS` sounds good.

Re: slicing it out of `.env` - I noticed that our handling of this is inconsistent. `CYPRESS_PROJECT_ID` and `CYPRESS_CRASH_REPORTS` is sliced out of `.env`, but `CYPRESS_RECORD_KEY` is not:

`CYPRESS_CRASH_REPORTS=0 CYPRESS_PROJECT_ID=testo CYPRESS_RECORD_KEY=fasa npm run cypress:open`

image

Do we also want to slice out `CYPRESS_RECORD_KEY` as part of this, or is that intentional?

**Comment by brian-mann** [comment]
Although it seems crazy at first - there's actually a few rules here and it's not **actually** inconsistent.

All `CYPRESS_*` env vars fall through **except** if they match a `cypress.json` property.

Because `CYPRESS_PROJECT_ID` matches the `projectId` config property, that's why it is suppressed from showing up as an environment variable.

I've gone ahead and let the `CYPRESS_CRASH_REPORTS` show up in the env vars to keep it consistent. This will all get refactored in the new proposal for `cypress.js` so I'm not really that worried about.

FWIW I'm also hiding record keys and truncating them. I'm opening a separate issue which displays `projectId` in the resolved configuration - since it was not added as a default `configKey`.

**Comment by YOU54F** [comment]
Thank you all for you quick and hard work on this 🙌 it’s a testament to your dedication to supporting OSS.

**Pull request merged**